/**:
  ros__parameters:
    linkage_num: 3

    # MARK: Links

    # === LINKAGE GEOMETRY === #
    # All dimensions in meters unless otherwise noted

    # Steering arm (link connecting wheel to tie rod)
    steering_arm_length: 0.150        # [m] link 'a' in kinematic calc
    steering_arm_width: 0.0254        # [m] visual dimension
    steering_arm_thickness: 0.0254    # [m] visual dimension

    # Tie rod (link connecting rack to steering arm)
    tie_rod_length: 0.1813            # [m] link 'b' in kinematic calc
    tie_rod_width: 0.0254             # [m] visual dimension
    tie_rod_thickness: 0.0254         # [m] visual dimension

    # Rack (slider mechanism)
    rack_length: 0.1500               # [m] total rack length
    rack_width: 0.0254                # [m] visual dimension
    rack_thickness: 0.0254            # [m] visual dimension
    rack_offset_x: -0.100             # [m] link 'c' in kinematic calc
    rack_neutral_y: 0.2298            # [m] neutral position 'd' in kinematic calc
    rack_travel_limit: 0.088          # [m] +/- travel from neutral

    # Pinion (rotary to linear converter)
    pinion_radius: 0.01905            # [m] 19.05 mm
    pinion_length: 0.0254             # [m] visual dimension
    pinion_offset_x: -0.12048          # [m] x position
    pinion_offset_z: 0.04445          # [m] z position
    pinion_angle_limit: 10         # [rad] +/- rotation limit
    pinion_gear_ratio: 1.652          # [unitless] gear ratio for pinion movement

    # Frame (main structural element)
    frame_length: 0.0254              # [m] visual dimension
    frame_width: 0.6096               # [m] 24 inches
    frame_thickness: 0.0254           # [m] visual dimension

    # Wheel positions (for visualization)
    wheel_radius: 0.1524              # [m] 6 inches
    wheel_length: 0.0762              # [m] 3 inches
    wheel_offset_x: 0.1               # [m] forward from steering arm
    wheel_angle:  0.32253             # [rad] correct for steering arm angle

    # Differential/rear axle
    diff_offset_x: -0.9144            # [m] -36 inches from base_link
    rear_wheel_spacing: 0.4048        # [m] +/- from diff center

    # Chassis (for visualization in linkage model)
    chassis_length: 0.9144            # [m] 36 inches
    chassis_width: 0.6096             # [m] 24 inches
    chassis_height: 0.1524            # [m] 6 inches
    chassis_offset_x: -0.4572         # [m]
    chassis_offset_z: 0.0762          # [m]

    # === VEHICLE KINEMATICS === #
    wheel_base: 0.6096                # [m] distance between front and rear axles (36 inches)
    track_width: 0.9144               # [m] distance between front and rear axles (36 inches)
    max_pinion_angle: 2.6             # [rad]
    max_steering_angle: 5.95           # [rad]

# MARK: Ctrl
controller_manager:
  ros__parameters:
    use_sim_time: true
    update_rate: 100 # Hz

    ackermann_angle_controller:
      type: igvc_control/AckermannAngleController

ackermann_angle_controller:
  ros__parameters:
      front_left_steering_joint: front_left_steering_joint
      front_right_steering_joint: front_right_steering_joint
      pinion_joint: pinion_joint
      back_left_wheel_joint: back_left_wheel_joint
      back_right_wheel_joint: back_right_wheel_joint
      rear_motor_joint: rear_motor_joint
      cmd_vel_topic: /cmd_vel
      reference_timeout: 2.0
      calibration_sample_size: 1024


# MARK: Joy
joy_node:
  ros__parameters:
    device_id: 0
    deadzone: 0.05
    autorepeat_rate: 20.0

teleop_node:
  ros__parameters:
    axis_linear:
      x: 1                            # linear velocity
    scale_linear:
      x: 2.3                          # max linear velocity (2.3 m/s)

    axis_angular:
      yaw: 0                          # yaw angular velocity
    scale_angular:
      yaw: 5.0                        # max yaw angular velocity (1.5 rad/s)

    require_enable_button: false



